src/software/simulation/physics/physics_world.cpp

```diff
@@ -1,78 +1,186 @@
-        TimestampedBallState timestamped_ball_state(physics_ball->getBallState(),
-                                                    current_timestamp);
-        new_world.mutableBall() = Ball(timestamped_ball_state);
+        if (isRobotIdAvailable(state_with_id.id, TeamColour::YELLOW))
+        {
+            yellow_physics_robots.emplace_back(std::make_shared<PhysicsRobot>(
+                state_with_id.id, b2_world, state_with_id.robot_state,
+                ROBOT_WITH_BATTERY_MASS_KG));
+        }
+        else
+        {
+            std::stringstream ss;
+            ss << "Yellow robot with id " << state_with_id.id
+               << " already exists in the physics world" << std::endl;
+            throw std::runtime_error(ss.str());
+        }
     }
-    if (physics_field)
+}
+
+void PhysicsWorld::addBlueRobots(const std::vector<RobotStateWithId>& robots)
+{
+    for (const auto& state_with_id : robots)
     {
-        new_world.mutableField() = physics_field->getField();
+        if (isRobotIdAvailable(state_with_id.id, TeamColour::BLUE))
+        {
+            blue_physics_robots.emplace_back(std::make_shared<PhysicsRobot>(
+                state_with_id.id, b2_world, state_with_id.robot_state,
+                ROBOT_WITH_BATTERY_MASS_KG));
+        }
+        else
+        {
+            std::stringstream ss;
+            ss << "Blue robot with id " << state_with_id.id
+               << " already exists in the physics world" << std::endl;
+            throw std::runtime_error(ss.str());
+        }
     }
+}
 
-    new_world.mutableFriendlyTeam().clearAllRobots();
-    std::vector<Robot> friendly_robots;
-    for (const auto& robot : friendly_physics_robots)
+const RobotId PhysicsWorld::getAvailableRobotId(TeamColour colour) const
+{
+    for (RobotId i = 0; i < std::numeric_limits<RobotId>::max(); i++)
     {
-        TimestampedRobotState timestamped_robot_state(robot->getRobotState(),
-                                                      current_timestamp);
-        friendly_robots.emplace_back(Robot(robot->getRobotId(), timestamped_robot_state));
+        if (isRobotIdAvailable(i, colour))
+        {
+            return i;
+        }
     }
-    new_world.mutableFriendlyTeam().updateRobots(friendly_robots);
 
-    new_world.mutableEnemyTeam().clearAllRobots();
-    std::vector<Robot> enemy_robots;
-    for (const auto& robot : enemy_physics_robots)
+    if (isRobotIdAvailable(std::numeric_limits<RobotId>::max(), colour))
     {
-        TimestampedRobotState timestamped_robot_state(robot->getRobotState(),
-                                                      current_timestamp);
-        enemy_robots.emplace_back(Robot(robot->getRobotId(), timestamped_robot_state));
+        return std::numeric_limits<RobotId>::max();
     }
-    new_world.mutableEnemyTeam().updateRobots(enemy_robots);
 
-    return new_world;
+    LOG(FATAL) << "Out of available " << toString(colour)
+               << " robot IDs in the physics world";
```

Comment:
As a general comment, it's worth thinking about if we ever really want to use `LOG(FATAL)`, as I don't think it throws an exception (I think it effectively kills the program), which is not desirable in a gameplay scenario. 

Let me know when you see this and we can bat it back and forth a bit in person.

Reply:
Opened https://github.com/UBC-Thunderbots/Software/issues/1430 for this, just as a note.

Reply:
Will save this to be addressed in #1430. This is a pretty unrecoverable (and as has been pointed out, impossible) situation so I think it's reasonable here.